pkg/distributor/distributor.go

```diff
@@ -781,7 +786,8 @@ func (d *Distributor) wrapPushWithMiddlewares(next PushFunc) PushFunc {
 	// The middlewares will be applied to the request (!) in the specified order, from first to last.
 	// To guarantee that, middleware functions will be called in reversed order, wrapping the
 	// result from previous call.
-	middlewares = append(middlewares, d.limitsMiddleware) // should run first because it checks limits before other middlewares need to read the request body
+	middlewares = append(middlewares, d.checkStartedMiddleware) // This middleware doesn't read the request body.
```

Comment:
Is there a reason not to put this middleware after the limits middleware?

Reply:
Or *inside* limits middleware? It's just one `if` check.

(Node that even if middleware doesn't read the request body, it doesn't mean that request body isn't already in memory. This is especially true when receiving request via httpgrpc -- entire request is in memory, just not decompressed yet.)

Reply:
I thought it made sense in front of the limits middleware because the limits middleware depends on runtime config to load distributor instance limits; runtime config is one of the services that might not have been properly initialized before the distributor begins handling push requests.

Looking more closely at the limits middleware though, I'm thinking we could still put this check after the rate limiting portion but before the request size check. I think it'd also make sense to change to rejecting instead of blocking, heeding your comment below:

> Should we block push request, or reject it? By blocking here, we can pile up enough requests in the memory that distributor can already crash.

Reply:
I think I was wrong about where we're dependent on runtime config; I've moved the check to the beginning of the limits middleware in https://github.com/grafana/mimir/pull/9317/commits/010354af2db1e43763209e4c0ce8689a8c129991